File: identity/src/main/java/com/android/identity/DataTransportOptions.java
package com.android.identity;

import androidx.annotation.NonNull;

/**
 * A set of options used when creating a {@link DataTransport} derived instance.
 */
public class DataTransportOptions {

    private boolean mBleUseL2CAP;
    private boolean mBleClearCache;

    DataTransportOptions() {}

    /**
     * Returns the preference for use BLE L2CAP transmission profile.
     *
     * <p>If true, L2CAP will be used if supported by the OS and remote mdoc.
     *
     * @return indicates if L2CAP should be used, if available.
     */
    public boolean getBleUseL2CAP() {
        return mBleUseL2CAP;
    }

    /**
     * Returns the preference to clear the BLE Service Cache before service discovery when acting as
     * a GATT Client.
     *
     * @return indicates if the BLE Service Cache should be cleared.
     */
    public boolean getBleClearCache() {
        return mBleClearCache;
    }

    /**
     * A builder for {@link DataTransportOptions}.
     */
    public static class Builder {
        DataTransportOptions mOptions;

        /**
         * Creates a new builder.
         */
        public Builder() {
            mOptions = new DataTransportOptions();
        }

        /**
         * Sets the preference for use BLE L2CAP transmission profile.
         *
         * <p>Use L2CAP if supported by the OS and remote mdoc.
         *
         * <p>The default value for this is <em>false</em>.
         *
         * @param useL2CAP indicates if it should use L2CAP socket if available.
         * @return the builder.
         */
        public Builder setBleUseL2CAP(boolean useL2CAP) {
            mOptions.mBleUseL2CAP = useL2CAP;
            return this;
        }

        /**
         * Sets whether to clear the BLE Service Cache before service discovery when acting as
         * a GATT Client.
         *
         * <p>The default value for this is <em>false</em>.
         *
         * @param bleClearCache indicates if the BLE Service Cache should be cleared.
         * @return the builder.
         */
        public Builder setBleClearCache(boolean bleClearCache) {
            mOptions.mBleClearCache = bleClearCache;
            return this;
        }

        /**
         * Builds the {@link DataTransportOptions}.
         *
         * @return the built {@link DataTransportOptions} instance.
         */
        public @NonNull DataTransportOptions build() {
            return mOptions;
        }
    }
}

File: identity-android/src/main/java/com/android/identity/android/mdoc/transport/DataTransportOptions.kt
package com.android.identity.android.mdoc.transport

/**
 * A set of options used when creating a [DataTransport] derived instance.
 */
class DataTransportOptions internal constructor(
    /**
     * The preference for using BLE L2CAP transmission profile.
     *
     * If true, L2CAP will be used if supported by the OS and remote mdoc.
     */
    val bleUseL2CAP: Boolean,

    /**
     * The preference to clear the BLE Service Cache before service discovery when
     * acting as a GATT Client.
     */
    val bleClearCache: Boolean,

    /**
     * Whether BLE L2CAP PSM is conveyed in engagement.
     *
     * See [Builder.setExperimentalBleL2CAPPsmInEngagement] for details on this option.
     */
    val experimentalBleL2CAPPsmInEngagement: Boolean,
) {
    /**
     * A builder for [DataTransportOptions].
     */
    class Builder {
        private var bleUseL2CAP = false
        private var bleClearCache = false
        private var experimentalBleL2CAPPsmInEngagement = false

        /**
         * Sets the preference for use BLE L2CAP transmission profile.
         *
         * Use L2CAP if supported by the OS and remote mdoc.
         *
         * The default value for this is `false`.
         *
         * @param value indicates if it should use L2CAP socket if available.
         * @return the builder.
         */
        fun setBleUseL2CAP(value: Boolean) = apply {
            bleUseL2CAP = value
        }

        /**
         * Sets whether to clear the BLE Service Cache before service discovery when acting as
         * a GATT Client.
         *
         * The default value for this is `false`.
         *
         * @param value indicates if the BLE Service Cache should be cleared.
         * @return the builder.
         */
        fun setBleClearCache(value: Boolean) = apply {
            bleClearCache = value
        }

        /**
         * Sets whether the BLE L2CAP PSM is conveyed in the engagement.
         *
         * This uses a non-standardized mechanisms for conveying the BLE L2CAP PSM
         * in NFC and QR engagement.
         *
         * The default value for this is *false*.
         *
         * @param value
         * @return the builder.
         */
        fun setExperimentalBleL2CAPPsmInEngagement(value: Boolean) = apply {
            experimentalBleL2CAPPsmInEngagement = value
        }

        /**
         * Builds the [DataTransportOptions].
         *
         * @return the built [DataTransportOptions] instance.
         */
        fun build(): DataTransportOptions {
            return DataTransportOptions(
                bleUseL2CAP,
                bleClearCache,
                experimentalBleL2CAPPsmInEngagement
            )
        }
    }
}
